baremetal/metal3machine_manager.go

```diff
@@ -809,20 +899,56 @@ func (m *MachineManager) chooseHost(ctx context.Context) (*bmh.BareMetalHost, *p
 		}
 
 		if labelSelector.Matches(labels.Set(host.ObjectMeta.Labels)) {
-			m.Log.Info("Host matched hostSelector for Metal3Machine", "host", host.Name)
-			availableHosts = append(availableHosts, &hosts.Items[i])
+			if m.nodeReuseLabelExists(ctx, &host) && m.nodeReuseLabelMatches(ctx, &host) {
+				m.Log.Info("Found host with matching nodeReuseLabelName", "host", host.Name)
+				availableHostsWithNodeReuse = append(availableHostsWithNodeReuse, &hosts.Items[i])
+			} else if !m.nodeReuseLabelExists(ctx, &host) {
+				m.Log.Info("Host matched hostSelector for Metal3Machine", "host", host.Name)
+				availableHosts = append(availableHosts, &hosts.Items[i])
+			}
 		} else {
 			m.Log.Info("Host did not match hostSelector for Metal3Machine", "host", host.Name)
 		}
 	}
+
+	m.Log.Info(fmt.Sprintf("%d hosts available with nodeReuseLabelName while choosing host for Metal3 machine", len(availableHostsWithNodeReuse)))
 	m.Log.Info(fmt.Sprintf("%d hosts available while choosing host for Metal3 machine", len(availableHosts)))
-	if len(availableHosts) == 0 {
+	if len(availableHostsWithNodeReuse) == 0 && len(availableHosts) == 0 {
 		return nil, nil, nil
 	}
 
-	// choose a host at random from available hosts
+	// choose a host
 	rand.Seed(time.Now().Unix())
-	chosenHost := availableHosts[rand.Intn(len(availableHosts))]
+	var chosenHost *bmh.BareMetalHost
+
+	// If there are hosts with nodeReuseLabelName:
+	if len(availableHostsWithNodeReuse) != 0 {
+		for _, host := range availableHostsWithNodeReuse {
+			// Build list of hosts in Ready state with nodeReuseLabelName
+			hostsInReadyStateWithNodeReuse := []*bmh.BareMetalHost{}
+			// Build list of hosts in any other state than Ready state with nodeReuseLabelName
+			hostsInNotReadyStateWithNodeReuse := []*bmh.BareMetalHost{}
+			if host.Status.Provisioning.State == bmh.StateReady {
+				hostsInReadyStateWithNodeReuse = append(hostsInReadyStateWithNodeReuse, host)
+			} else {
+				hostsInNotReadyStateWithNodeReuse = append(hostsInNotReadyStateWithNodeReuse, host)
+			}
+
+			// If host is found in `Ready` state, pick it
+			if len(hostsInReadyStateWithNodeReuse) != 0 {
+				m.Log.Info(fmt.Sprintf("Found %v host(s) with nodeReuseLabelName in Ready state", len(hostsInReadyStateWithNodeReuse)))
+				chosenHost = hostsInReadyStateWithNodeReuse[rand.Intn(len(hostsInReadyStateWithNodeReuse))]
+			} else if len(hostsInNotReadyStateWithNodeReuse) != 0 {
+				m.Log.Info(fmt.Sprintf("Found %v host(s) with nodeReuseLabelName in other state than Ready, requeuing", len(hostsInNotReadyStateWithNodeReuse)))
+				return nil, nil, &RequeueAfterError{RequeueAfter: requeueAfter}
+			}
+		}
```

Comment:
@namnx228 thanks for the review. If we reached this part of the code, it means we are only iterating over `availableHostsWithNodeReuse` which means only hosts with node reuse label. In that case, choosing the first host found in 'ready' state is fine.

Reply:
That makes sense. If you always choose the first node in the availableHostsWithNodeReuse list, you can pick hostsInReadyStateWithNodeReuse[0] since calling the random function is not necessary here.

Reply:
Agree, choosing the first host from the list with index is an alternative, but there is no strict requirement to choose hosts by index in case if there are more than one. Also, `rand` was used to be consistent with the else condition of `chosenHost`  list.